lightningd/jsonrpc.c

```diff
@@ -575,14 +578,140 @@ struct json_stream *json_stream_fail(struct command *cmd,
 	return r;
 }
 
+static struct command_result *command_exec(struct json_connection *jcon,
+                                           struct command *cmd,
+                                           const char *buffer,
+                                           const jsmntok_t *request,
+                                           const jsmntok_t *params)
+{
+	struct command_result *res;
+
+	res = cmd->json_cmd->dispatch(cmd, buffer, request, params);
+
+	assert(res == &param_failed
+	       || res == &complete
+	       || res == &pending
+	       || res == &unknown);
+
+	/* If they didn't complete it, they must call command_still_pending.
+	 * If they completed it, it's freed already. */
+	if (res == &pending)
+		assert(cmd->pending);
+
+	list_for_each(&jcon->commands, cmd, list)
+		assert(cmd->pending);
+
+	return res;
+}
+
+/* A plugin hook to take over (fail/alter) RPC commands */
+struct rpc_command_hook_payload {
+	struct command *cmd;
+	const char *buffer;
+	const jsmntok_t *request;
+};
+
+static void rpc_command_hook_serialize(struct rpc_command_hook_payload *p,
+                                       struct json_stream *s)
+{
+	const jsmntok_t *method, *params;
+
+	method = json_get_member(p->buffer, p->request, "method");
+	params = json_get_member(p->buffer, p->request, "params");
+
+	json_object_start(s, "rpc_command");
+	json_add_string(s, "method", tal_fmt(tmpctx, "%.*s",
+	                                     method->end - method->start,
+	                                     p->buffer + method->start));
+	json_add_string(s, "params", tal_fmt(tmpctx, "%.*s",
```

Comment:
Converting the params to a string, so they need to be unpacked on the plugin side is a bit of a strange choice. Why not use `json_add_tok` to create a direct copy instead?

Reply:
> is a bit of a strange choice

Hmmm it is ^^ I'll give `json_add_tok` a shot.

Reply:
Ok that's far better using `json_add_tok` ! An excuse is that i wrote this helper after writing the serialization function..